### ISSUE TITLE
Language server error: Python UNABLE_TO_GET_ISSUER_CERT_LOCALLY

### ISSUE THREAD

**Issue by ericbn** [zed-industries/community#1290]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Hi 👋 

After starting Zed, it shows in the status bar:

> Checking for updates to Python language server...

Then:

> Failed to download Python language server. Click to show error.

After clicking on the message, I get a buffer containing:
```
Language server error: Python

failed to execute npm info:
stdout: "{\n  \"error\": {\n    \"code\": \"UNABLE_TO_GET_ISSUER_CERT_LOCALLY\",\n    \"summary\": \"request to https://registry.npmjs.org/pyright failed, reason: unable to get local issuer certificate\",\n    \"detail\": \"\"\n  }\n}\n"
stderr: "npm ERR! code UNABLE_TO_GET_ISSUER_CERT_LOCALLY\nnpm ERR! errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY\nnpm ERR! request to https://registry.npmjs.org/pyright failed, reason: unable to get local issuer certificate\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /Users/me/.npm/_logs/2023-03-24T02_09_47_304Z-debug-0.log\n"
```
The contents of the log file mentioned in the message above is:
```
0 verbose cli /usr/local/Cellar/node/19.8.1/bin/node /usr/local/bin/npm
1 info using npm@9.5.1
2 info using node@v19.8.1
3 timing npm:load:whichnode Completed in 1ms
4 timing config:load:defaults Completed in 1ms
5 timing config:load:file:/usr/local/lib/node_modules/npm/npmrc Completed in 2ms
6 timing config:load:builtin Completed in 2ms
7 timing config:load:cli Completed in 2ms
8 timing config:load:env Completed in 0ms
9 timing config:load:file:/.npmrc Completed in 0ms
10 timing config:load:project Completed in 1ms
11 timing config:load:file:/Users/me/.npmrc Completed in 0ms
12 timing config:load:user Completed in 1ms
13 timing config:load:file:/usr/local/etc/npmrc Completed in 0ms
14 timing config:load:global Completed in 0ms
15 timing config:load:setEnvs Completed in 1ms
16 timing config:load Completed in 8ms
17 timing npm:load:configload Completed in 9ms
18 timing npm:load:mkdirpcache Completed in 0ms
19 timing npm:load:mkdirplogs Completed in 0ms
20 verbose title npm info pyright
21 verbose argv "-fetch-retry-mintimeout" "2000" "-fetch-retry-maxtimeout" "5000" "-fetch-timeout" "5000" "info" "pyright" "--json"
22 timing npm:load:setTitle Completed in 4ms
23 timing config:load:flatten Completed in 3ms
24 timing npm:load:display Completed in 4ms
25 verbose logfile logs-max:10 dir:/Users/me/.npm/_logs/2023-03-24T02_09_47_304Z-
26 verbose logfile /Users/me/.npm/_logs/2023-03-24T02_09_47_304Z-debug-0.log
27 timing npm:load:logFile Completed in 5ms
28 timing npm:load:timers Completed in 0ms
29 timing npm:load:configScope Completed in 0ms
30 timing npm:load Completed in 23ms
31 silly logfile start cleaning logs, removing 1 files
32 silly logfile done cleaning log files
33 timing command:info Completed in 7279ms
34 verbose type system
35 verbose stack FetchError: request to https://registry.npmjs.org/pyright failed, reason: unable to get local issuer certificate
35 verbose stack     at ClientRequest.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/minipass-fetch/lib/index.js:130:14)
35 verbose stack     at ClientRequest.emit (node:events:512:28)
35 verbose stack     at TLSSocket.socketErrorListener (node:_http_client:495:9)
35 verbose stack     at TLSSocket.emit (node:events:524:35)
35 verbose stack     at emitErrorNT (node:internal/streams/destroy:151:8)
35 verbose stack     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
35 verbose stack     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
36 verbose cwd /
37 verbose Darwin 21.6.0
38 verbose node v19.8.1
39 verbose npm  v9.5.1
40 error code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
41 error errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
42 error request to https://registry.npmjs.org/pyright failed, reason: unable to get local issuer certificate
43 verbose exit 1
44 timing npm Completed in 7319ms
45 verbose code 1
46 error A complete log of this run can be found in:
46 error     /Users/me/.npm/_logs/2023-03-24T02_09_47_304Z-debug-0.log
```
Here's the contents of my `~/Library/Application Support/Zed/languages` directory:

```
drwxr-xr-x   - me 23 Mar 21:08 .
drwxr-xr-x   - me 16 Mar 15:47 ├── json-language-server
.rwxr-xr-x 51M me 16 Mar 15:47 │  └── json-language-server-v0.3.0-x86_64
drwxr-xr-x   - me 23 Mar 21:08 └── pyright
```
Already tried removing the `pyright` subdirectory and restarting Zed and the error persists.

EDIT: For more context, I'm not behind a firewall or VPN, except for the macOS Firewall. I don't see any rule for Zed in the macOS Firewall. Also, Zed was able to check for updates, so I'm assuming it has network access. Also, running `curl -v https://registry.npmjs.org/pyright`  on the Zed terminal itself, gives me:
```
*   Trying 104.16.18.35:443...
* Connected to registry.npmjs.org (104.16.18.35) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-ECDSA-CHACHA20-POLY1305
* ALPN, server accepted to use h2
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=Cloudflare, Inc.; CN=sni.cloudflaressl.com
*  start date: Jul 18 00:00:00 2022 GMT
*  expire date: Jul 18 23:59:59 2023 GMT
*  subjectAltName: host "registry.npmjs.org" matched cert's "*.npmjs.org"
*  issuer: C=US; O=Cloudflare, Inc.; CN=Cloudflare Inc ECC CA-3
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7ff41d812c00)
> GET /pyright HTTP/2
> Host: registry.npmjs.org
> user-agent: curl/7.79.1
> accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 256)!
< HTTP/2 200
< date: Fri, 24 Mar 2023 02:23:41 GMT
< content-type: application/json
< content-length: 1133028
< cf-ray: 7acb741e7a213baf-GYE
< accept-ranges: bytes
< access-control-allow-origin: *
< age: 49
< cache-control: public, max-age=300
< etag: "3c5880a5bfe05110c52946ab77a2e2ce"
< last-modified: Wed, 22 Mar 2023 04:47:45 GMT
< vary: accept-encoding, accept
< cf-cache-status: HIT
< x-amz-replication-status: COMPLETED
< x-amz-server-side-encryption: AES256
< server: cloudflare
<
...
```

### Environment

Zed: v0.78.1 (stable)
OS: macOS 12.6.3
Memory: 16 GiB
Architecture: x86_64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

```
2023-03-24T02:09:45 [INFO] ========== starting zed ==========
2023-03-24T02:09:46 [INFO] Opening main db
2023-03-24T02:09:46 [INFO] set environment variables from shell:/usr/local/bin/zsh, path:/Users/me/perl5/bin:/usr/local/bin:/usr/local/sbin:/Users/me/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/opt/fzf/bin
2023-03-24T02:09:46 [INFO] Opening main db
2023-03-24T02:09:46 [INFO] open paths ["/Users/me/Developer/myproject"]
2023-03-24T02:09:46 [INFO] Opening main db
2023-03-24T02:09:46 [INFO] Opening main db
2023-03-24T02:09:46 [ERROR] No path stored for this editor
2023-03-24T02:09:54 [ERROR] no cached binary
2023-03-24T02:09:54 [ERROR] failed to execute npm info:
stdout: "{\n  \"error\": {\n    \"code\": \"UNABLE_TO_GET_ISSUER_CERT_LOCALLY\",\n    \"summary\": \"request to https://registry.npmjs.org/pyright failed, reason: unable to get local issuer certificate\",\n    \"detail\": \"\"\n  }\n}\n"
stderr: "npm ERR! code UNABLE_TO_GET_ISSUER_CERT_LOCALLY\nnpm ERR! errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY\nnpm ERR! request to https://registry.npmjs.org/pyright failed, reason: unable to get local issuer certificate\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /Users/me/.npm/_logs/2023-03-24T02_09_47_304Z-debug-0.log\n"
```

**Comment by diimdeep** [comment]
Might be related https://github.com/zed-industries/community/issues/1265

**Comment by ericbn** [comment]
This fixed the issue for me: https://github.com/Homebrew/homebrew-core/issues/92271

It was an issue with the homebrew-installed openssl@1.1, which is used by the homebrew-installed node.

I normally don't use the homebrew-installed node. How can we configure the path of the node executable we want to use with Zed?

**Comment by JosephTLyons** [comment]
> This fixed the issue for me: [Homebrew/homebrew-core#92271](https://github.com/Homebrew/homebrew-core/issues/92271)
> 
> It was an issue with the homebrew-installed openssl@1.1, which is used by the homebrew-installed node.
> 
> I normally don't use the homebrew-installed node. How can we configure the path of the node executable we want to use with Zed?

This isn't possible yet.  Currently, we are looking at the globally-installed version.  We should be landing a fix soon that will bundle the correct node version with the language server, where we don't rely on the global version at all.

**Comment by JosephTLyons** [comment]
I'm going to close this out, in favor of #1265, since it seems to be the same core issue.  That being said, even after that fixes land, you won't be able to specify a path for what node executable to use.  If that is something you'd still like, even once we land the fix to bundle node with the language server, feel free to open a new issue for it.

**Comment by ericbn** [comment]
Cool, the solution of bundling node with Zed sounds as a much better solution than having the user configure with node should be used. I love that.
